### PR TITLE
Respect constraints on type class implementations

### LINKED ISSUE #144: `test` methods should respect constraints
For example, `Array a` satisfies the requirements of [Setoid][1] if and only if `a` satisfies the requirements of Setoid. `Z.Setoid.test` currently ignores this constraint:

```javascript
> var Z = require ('sanctuary-type-classes')

> var Useless = require ('sanctuary-useless')

> Z.Setoid.test (Useless)
false

> Z.Setoid.test ([Useless])
true
```

`Z.Setoid.test ([])` and `Z.Setoid.test ([1, 2, 3])` should evaluate `true` but `Z.Setoid.test ([Useless])` should evaluate `false`. The behaviour of other `test` methods should also change.

[1]: https://github.com/fantasyland/fantasy-land/#setoid

### PR DESCRIPTION
**Pull request by Avaq**
Closes #144

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -873,7 +885,7 @@
     return {};
   }
 
-  //  Object$prototype$equals :: StrMap a ~> StrMap a -> Boolean
+  //  Object$prototype$equals :: Setoid a => StrMap a ~> StrMap a -> Boolean
```
Comment (davidchambers) [inline index.js:888]: :thumbsup:

File: index.js [hunk 2]
```diff
@@ -291,17 +291,29 @@
...
-            return getPrototypeMethod (_name) (x) != null;
-          })
-        );
-      }
+      (function() {
+        var $seen = [];
+        return function(x) {
+          if ($seen.includes (x)) {
+            return true;
+          }
```
Comment (davidchambers) [inline index.js:299]:
```suggestion
          if ($seen.includes (x)) return true;
```

File: index.js [hunk 3]
```diff
@@ -291,17 +291,27 @@
...
-          }) &&
-          prototypeMethodNames.every (function(_name) {
-            return getPrototypeMethod (_name) (x) != null;
-          })
-        );
-      }
+      (function() {
+        var $seen = [];
+        return function(x) {
+//        if ($seen.includes (x)) return true;
```
Comment (davidchambers) [inline index.js]: I don't think this code path can be exercised without the changes in the following commit. I think `npm test` should pass for every new commit on `master`, so I commented out this line and uncommented it in the following commit.

File: index.js [hunk 4]
```diff
@@ -294,7 +294,7 @@
       (function() {
         var $seen = [];
         return function(x) {
-//        if ($seen.includes (x)) return true;
+          if ($seen.includes (x)) return true;
```
Comment (davidchambers) [inline index.js]: I uncommented this line as the code path can now be exercised. :)

**Comment by Avaq** [comment]
Doctests LGTM

**Review by davidchambers — APPROVED** [review]

**Pull request merged**